README.md
# spotify-random-album


ChangeLog.md
# Changelog for spotify-random-album

## Unreleased changes


src/Lib.hs
{-# LANGUAGE OverloadedStrings   #-}
{-# LANGUAGE RecordWildCards     #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TypeApplications    #-}
module Lib
  ( appMain
  )
where

import           Control.Lens
import           Control.Monad
import           Control.Monad.IO.Class
import           Control.Monad.Random
import           Data.Aeson
import qualified Data.Aeson                    as Aeson
import           Data.Aeson.Lens
import qualified Data.ByteString.Base64        as Base64
import qualified Data.ByteString.Char8         as S8
import qualified Data.ByteString.Lazy.Char8    as LBS
import           Data.Foldable
import qualified Data.List                     as List
import qualified Data.Map.Strict               as Map
import           Data.String
import qualified Data.Text                     as T
import qualified Data.Text.Lazy                as TL
import           Data.Traversable
import           Data.Vault.Lazy               as Vault
import           GHC.Generics
import           Network.HTTP.Simple
import           Network.HTTP.Types
import           Network.Wai                   (vault)
import           Network.Wai.Session
import           Network.Wai.Session.Map
import           System.Environment
import           System.Random
import qualified Text.Blaze.Html.Renderer.Text as Blaze
import           Text.Blaze.Html5              (style, toHtml, (!))
import qualified Text.Blaze.Html5              as Blaze hiding (style)
import qualified Text.Blaze.Html5.Attributes   as Blaze hiding (form, label,
                                                         span)
import           Types.Spotify
import           Web.Cookie                    (defaultSetCookie)
import           Web.Scotty
import qualified Web.Scotty                    as Scotty

import           Html.Header
import qualified Html.Header                   as Header
import           Html.Layout
import           Html.RandomizeForm

-- Session keys
stateKey = "spotify_auth_state"
accessTokenKey = "access_token"
refreshTokenKey = "refresh_token"
sessionKey = "session"
albumIdsKey = "album_ids"

appMain :: IO ()
appMain = do
  clientId     <- getEnv "SPOTIFY_CLIENT_ID"
  clientSecret <- getEnv "SPOTIFY_CLIENT_SECRET"
  port         <- read <$> getEnv "PORT"
  redirectUri  <- S8.pack <$> getEnv "REDIRECT_URI"
  session      <- Vault.newKey
  store        <- mapStore_ @String

  let token = Base64.encode (S8.pack clientId <> ":" <> S8.pack clientSecret)
      authorization = "Basic " <> token
      sessionMiddleware = withSession store sessionKey defaultSetCookie session

  scotty port $ do
    Scotty.middleware sessionMiddleware

    Scotty.get "/" $ do
      renderPage $ do
        Blaze.h1 "Spotify Randomizer"
        Blaze.h3 "Play a random album from your Spotify collection"
        Blaze.a ! Blaze.class_ "spotify-login-button" ! Blaze.href "/login" $ do
          Blaze.i ! Blaze.class_ "fab fa-spotify" $ Blaze.text ""
          Blaze.text " Login with Spotify"

    Scotty.get "/login" $ do
      state <- liftIO $ fmap S8.pack $ replicateM 16 $ randomRIO ('a', 'z')

      -- Store state in a cookie
      sessionInsert session stateKey state

      let
        url =
          TL.pack
            $ S8.unpack
            $ ("https://accounts.spotify.com/authorize?" <>)
            $ mconcat
                [ "response_type=code"
                , "&client_id=" <> S8.pack clientId
                , "&scope=user-read-private user-read-email user-read-playback-state user-modify-playback-state user-library-read"
                , "&redirect_uri=" <> redirectUri
                , "&state=" <> state
                ]
      redirect url

    Scotty.get "/callback" $ do
      code  <- param "code"

      -- Store state in a cookie
      state <- param "state"
      mStoredState :: Maybe S8.ByteString <- sessionLookup session stateKey
      when (Just state /= mStoredState) $ error "mismatched state"

      request0 <- parseRequest "POST https://accounts.spotify.com/api/token"

      let request =
            setRequestBodyURLEncoded
                [ ("grant_type"  , "authorization_code")
                , ("code"        , code)
                , ("redirect_uri", redirectUri)
                ]
              $ setRequestHeader "Authorization" [authorization] request0

      tokenResponse <- httpJSON request

      let
        tokenResponseValue = getResponseBody @Value tokenResponse
        Just accessToken   = tokenResponseValue ^? key "access_token" . _String
        Just refreshToken  = tokenResponseValue ^? key "refresh_token" . _String

      sessionInsert session accessTokenKey  (S8.pack $ T.unpack accessToken)
      sessionInsert session refreshTokenKey (S8.pack $ T.unpack refreshToken)
      redirect "/choose"

    Scotty.get "/choose" $ do

      mAccessTokenBS <- sessionLookup session accessTokenKey
      let mAccessToken = fmap (T.pack . S8.unpack) mAccessTokenBS
      case mAccessToken of
        Nothing          -> redirect "/login"
        Just accessToken -> do
          devices@SpotifyDevices {..} <- getDevices accessToken
          case spotifyDevicesDevices of
            [] -> renderPage $ do
              Blaze.p $ mconcat
                [ "No Spotify Devices Found."
                , " Make sure that Spotify is active on at least one device and refresh this page"
                , " (sometimes, you'll need to start playing a song to mark the device as active)."
                ]
            _ -> renderPage $ randomizeForm devices

    Scotty.get "/randomize" $ do
      deviceId       <- param @S8.ByteString "device_id"
      mAccessTokenBS <- sessionLookup session accessTokenKey
      let mAccessToken = fmap (T.pack . S8.unpack) mAccessTokenBS
      case mAccessToken of
        Nothing          -> redirect "/login"
        Just accessToken -> do
          album <- do
            mAlbumIds <- sessionLookup session albumIdsKey
            case mAlbumIds of
              Just albumIdsStr -> liftIO $ do
                let albumIds = read @[String] $ S8.unpack albumIdsStr
                getAlbum accessToken =<< uniform albumIds

              Nothing -> do
                albums <- liftIO $ getMyAlbums accessToken
                let albumIds =
                      map (spotifyAlbumId . spotifyAlbumItemAlbum) albums
                sessionInsert session albumIdsKey (fromString $ show albumIds)
                liftIO $ uniform (map spotifyAlbumItemAlbum albums)
          let tracks = spotifyTracksItems . spotifyAlbumTracks $ album
              mSpotifyUrl =
                Map.lookup "spotify" (spotifyAlbumExternalUrls album)
              uris = map spotifyTrackUri tracks
              ids  = map spotifyTrackId tracks

          liftIO
            $ callSpotifyWithNoResponse
                accessToken
                "PUT https://api.spotify.com/v1/me/player/play"
            $ setRequestBodyJSON (object ["uris" Aeson..= uris])
            . setRequestQueryString [("device_id", Just deviceId)]

          devices <- getDevices accessToken

          let imageUrl = case spotifyAlbumImages album of
                []                      -> "#"
                (SpotifyImage {..} : _) -> spotifyImageUrl
              albumTitle = case mSpotifyUrl of
                Nothing -> Blaze.a
                Just spotifyUrl ->
                  (Blaze.a ! Blaze.href (fromString spotifyUrl) ! Blaze.target
                    "_blank"
                  )

          renderPage $ do
            Blaze.div ! Blaze.style "margin-left:auto; margin-right:auto" $ do
              Blaze.h2 $ do
                albumTitle (fromString $ spotifyAlbumName album)
                fromString
                  $  " ("
                  <> (take 4 $ spotifyAlbumReleaseDate album)
                  <> ")"

              Blaze.h3
                ( fromString
                $ List.intercalate ", "
                $ map spotifyArtistSimpleName
                $ spotifyAlbumArtists album
                )

              Blaze.h2
                (fromString $ List.intercalate ", " $ spotifyAlbumGenres album)

              Blaze.img
                ! Blaze.class_ "album-art"
                ! Blaze.src (fromString imageUrl)
                ! Blaze.style "max-height:440px"

              randomizeForm devices

renderFeature albumAudioFeatures name feature =
  Blaze.p $ fromString $ (name <> ": ") <> formatAsPercentage
    (averageFeature feature albumAudioFeatures)

formatAsPercentage d = show (floor (d * 100)) <> "%"

getMyAlbums accessToken = go [] "https://api.spotify.com/v1/me/albums?limit=50"
 where
  go albums url = do
    SpotifyAlbumsResponse {..} <- callSpotifyWith accessToken ("GET " <> url) id
    case spotifyAlbumsResponseNext of
      Nothing   -> pure (spotifyAlbumsResponseItems ++ albums)
      Just next -> (spotifyAlbumsResponseItems ++) <$> go albums next

getAlbum accessToken albumId = callSpotifyWith accessToken ("GET " <> url) id
  where url = "https://api.spotify.com/v1/albums/" <> albumId

averageFeature
  :: (SpotifyAudioFeature -> Double) -> SpotifyAudioFeatures -> Double
averageFeature feature (SpotifyAudioFeatures features) = average
  $ map feature features
  where average xs = sum xs / List.genericLength xs

callSpotifyWith
  :: forall a
   . (FromJSON a, Show a)
  => T.Text
  -> String
  -> (Request -> Request)
  -> IO a
callSpotifyWith accessToken request f = do
  request0 <- parseRequest request

  let spotifyAuthorization = "Bearer " <> S8.pack (T.unpack accessToken)
      request =
        setRequestHeader "Authorization" [spotifyAuthorization]
          $ setRequestHeader "Content-Type" ["application/json"]
          $ f request0

  response <- httpJSON request
  let responseBodyValue = getResponseBody @Value response
  if getResponseStatusCode response < 300
    then do
      case fromJSON @a responseBodyValue of
        Error str -> do
          print responseBodyValue
          error str
        Success a -> pure a
    else do
      print responseBodyValue
      error (show response)

callSpotifyWithNoResponse :: T.Text -> String -> (Request -> Request) -> IO ()
callSpotifyWithNoResponse accessToken request f = do
  request0 <- parseRequest request

  let spotifyAuthorization = "Bearer " <> S8.pack (T.unpack accessToken)
      request =
        setRequestHeader "Authorization" [spotifyAuthorization]
          $ setRequestHeader "Content-Type" ["application/json"]
          $ f request0

  response <- httpNoBody request
  if getResponseStatusCode response == 204
    then pure ()
    else do
      response1 <- httpJSON request
      print (getResponseBody response1 :: Value)
      error
        ("callSpotifyWithNoResponse Failure: " <> show response1 <> "\n" <> show
          (getResponseBody response1)
        )

sessionInsert session0 key val = do
  request0 <- request
  let mSessionAccess = Vault.lookup session0 (vault request0)
  case mSessionAccess of
    Nothing                  -> error "impossible: session is not available"
    Just (_, sessionInsert0) -> sessionInsert0 key val

sessionLookup session0 key = do
  request0 <- request
  let mSessionAccess = Vault.lookup session0 (vault request0)
  case mSessionAccess of
    Nothing                  -> pure Nothing
    Just (sessionLookup0, _) -> sessionLookup0 key

getDevices accessToken = liftIO $ callSpotifyWith
  accessToken
  "GET https://api.spotify.com/v1/me/player/devices"
  id

renderPage = Scotty.html . Blaze.renderHtml . withLayout


src/Html/Header.hs
{-# LANGUAGE NoImplicitPrelude #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE QuasiQuotes       #-}
module Html.Header where

import           Prelude                     (($))

import           Data.String
import           Text.Blaze.Html5
import qualified Text.Blaze.Html5.Attributes as A
import           Text.RawString.QQ

header = head $ do
  title "Spotify Randomizer"
  link
    ! A.href "https://fonts.googleapis.com/css?family=Montserrat&display=swap"
    ! A.rel "stylesheet"
  script ! A.src "https://kit.fontawesome.com/9b2f769989.js" ! customAttribute "crossorigin" "anonymous" $ do
    toHtml ("" :: String)
  style $ fromString appStyle

green = "#1DB954"
black = "#191414"

appStyle = [r|
body {
  font-family: 'Montserrat', sans-serif;
}

h3 {
  color: #191414;
  text-align: center;
  line-height: 0.5em;
}

h2 {
  text-align: center;
}

a {
  color: #1DB954;
  text-decoration: none;
}

.randomize-button {
  display: block;
  width: 100%;
  border: 3px solid #FFFFFF;
  color: #1DB954;
  padding: 14px 28px;
  font-size: 20px;
  cursor: pointer;
  text-align: center;
}

.container {
  text-align: center;
  width: 100%;
  position: relative;
  min-height: 100vh;
}

.content {
  display: inline-block;
  max-width:600px;
  padding-bottom: 2.5rem;    /* Footer height */
}

.spotify-login-button {
  background: #FFFFFF;
  margin: 14px;
  padding: 6px;
  border: 10px solid #FFFFFF;
  font-size: 24px;
}

footer {
  position: absolute;
  bottom: 0;
  width: 100%;
  height: 2.5rem;            /* Footer height */
  text-align: center;
}

.album-art {
  border: 5px solid #191414;
}
|]


src/Types/Spotify.hs
module Types.Spotify where

import           Data.Aeson
import qualified Data.Aeson      as Aeson
import           Data.Map.Strict (Map)
import           GHC.Generics

data SpotifyDevices = SpotifyDevices
  { spotifyDevicesDevices       :: [SpotifyDevice]
  } deriving (Show, Generic)

instance FromJSON SpotifyDevices where
  parseJSON = genericParseJSON (unPrefix "spotifyDevices")
instance ToJSON SpotifyDevices where
  toJSON = genericToJSON (unPrefix "spotifyDevices")
  toEncoding = genericToEncoding (unPrefix "spotifyDevices")

data SpotifyDevice = SpotifyDevice
  { spotifyDeviceId               :: String
  , spotifyDeviceIsActive         :: Bool
  , spotifyDeviceName             :: String
  , spotifyDeviceIsPrivateSession :: Bool
  , spotifyDeviceIsRestricted     :: Bool
  , spotifyDeviceVolumePercent    :: Double
  } deriving (Show, Generic)

instance FromJSON SpotifyDevice where
  parseJSON = genericParseJSON (unPrefix "spotifyDevice")

instance ToJSON SpotifyDevice where
  toJSON = genericToJSON (unPrefix "spotifyDevice")
  toEncoding = genericToEncoding (unPrefix "spotifyDevice")

data SpotifyAlbumsResponse = SpotifyAlbumsResponse
  { spotifyAlbumsResponseItems :: [SpotifyAlbumItem]
  , spotifyAlbumsResponseNext  :: Maybe String
  } deriving (Show, Generic)

instance ToJSON SpotifyAlbumsResponse where
  toJSON = genericToJSON $ unPrefix "spotifyAlbumsResponse"
instance FromJSON SpotifyAlbumsResponse where
  parseJSON = genericParseJSON $ unPrefix "spotifyAlbumsResponse"

data SpotifyAlbumItem = SpotifyAlbumItem
  { spotifyAlbumItemAlbum :: SpotifyAlbum
  } deriving (Show, Generic)

instance ToJSON SpotifyAlbumItem where
  toJSON = genericToJSON $ unPrefix "spotifyAlbumItem"
instance FromJSON SpotifyAlbumItem where
  parseJSON = genericParseJSON $ unPrefix "spotifyAlbumItem"

data SpotifyAlbum = SpotifyAlbum
  { spotifyAlbumUri          :: String
  , spotifyAlbumHref         :: String
  , spotifyAlbumId           :: String
  , spotifyAlbumExternalUrls :: Map String String
  , spotifyAlbumReleaseDate  :: String
  , spotifyAlbumGenres       :: [String]
  , spotifyAlbumName         :: String
  , spotifyAlbumTracks       :: SpotifyTracks
  , spotifyAlbumImages       :: [SpotifyImage]
  , spotifyAlbumArtists      :: [SpotifyArtistSimple]
  } deriving (Show, Generic)

instance ToJSON SpotifyAlbum where
  toJSON = genericToJSON $ unPrefix "spotifyAlbum"
instance FromJSON SpotifyAlbum where
  parseJSON = genericParseJSON $ unPrefix "spotifyAlbum"

data SpotifyImage = SpotifyImage
  { spotifyImageUrl :: String
  }
  deriving (Show, Generic)

instance ToJSON SpotifyImage where
  toJSON = genericToJSON $ unPrefix "spotifyImage"
instance FromJSON SpotifyImage where
  parseJSON = genericParseJSON $ unPrefix "spotifyImage"

data SpotifyArtistSimple = SpotifyArtistSimple
  { spotifyArtistSimpleName :: String
  }
  deriving (Show, Generic)

instance ToJSON SpotifyArtistSimple where
  toJSON = genericToJSON $ unPrefix "spotifyArtistSimple"
instance FromJSON SpotifyArtistSimple where
  parseJSON = genericParseJSON $ unPrefix "spotifyArtistSimple"

data SpotifyTracks = SpotifyTracks { spotifyTracksItems :: [SpotifyTrack], spotifyTracksNext  :: Maybe String }
  deriving (Show, Generic)

instance ToJSON SpotifyTracks where
  toJSON = genericToJSON $ unPrefix "spotifyTracks"
instance FromJSON SpotifyTracks where
  parseJSON = genericParseJSON $ unPrefix "spotifyTracks"

data SpotifyTrack = SpotifyTrack { spotifyTrackUri :: String, spotifyTrackId :: String }
  deriving (Show, Generic)

instance ToJSON SpotifyTrack where
  toJSON = genericToJSON $ unPrefix "spotifyTrack"
instance FromJSON SpotifyTrack where
  parseJSON = genericParseJSON $ unPrefix "spotifyTrack"

data SpotifyAudioFeatures = SpotifyAudioFeatures { spotifyAudioFeaturesAudioFeatures :: [SpotifyAudioFeature] }
  deriving (Show, Generic)

instance ToJSON SpotifyAudioFeatures where
  toJSON = genericToJSON $ unPrefix "spotifyAudioFeatures"
instance FromJSON SpotifyAudioFeatures where
  parseJSON = genericParseJSON $ unPrefix "spotifyAudioFeatures"

data SpotifyAudioFeature = SpotifyAudioFeature
  { spotifyAudioFeatureDanceability     :: Double
  , spotifyAudioFeatureEnergy           :: Double
  , spotifyAudioFeatureKey              :: Double
  , spotifyAudioFeatureLoudness         :: Double
  , spotifyAudioFeatureMode             :: Double
  , spotifyAudioFeatureSpeechiness      :: Double
  , spotifyAudioFeatureAcousticness     :: Double
  , spotifyAudioFeatureInstrumentalness :: Double
  , spotifyAudioFeatureLiveness         :: Double
  , spotifyAudioFeatureValence          :: Double
  , spotifyAudioFeatureTempo            :: Double
  , spotifyAudioFeatureId               :: String
  , spotifyAudioFeatureUri              :: String
  }
  deriving (Show, Generic)

instance ToJSON SpotifyAudioFeature where
  toJSON = genericToJSON $ unPrefix "spotifyAudioFeature"
instance FromJSON SpotifyAudioFeature where
  parseJSON = genericParseJSON $ unPrefix "spotifyAudioFeature"


unPrefix str = defaultOptions
  { fieldLabelModifier = camelTo2 '_' . drop (length (str :: String))
  }


src/Html/RandomizeForm.hs
{-# LANGUAGE NoImplicitPrelude #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE QuasiQuotes       #-}
{-# LANGUAGE RecordWildCards   #-}
module Html.RandomizeForm where

import           Prelude                     (($))
import qualified Prelude                     as P

import           Data.Foldable               (for_)
import           Data.String
import           Text.Blaze.Html5
import qualified Text.Blaze.Html5.Attributes as A
import           Text.RawString.QQ
import           Types.Spotify

randomizeForm SpotifyDevices {..} = (form ! A.action "/randomize") $ do
  let randomizeText = do
        i ! A.class_ "fas fa-random" $ empty
        toHtml (" Randomize" :: String)
  button randomizeText ! A.class_ "randomize-button"

  label ! A.for "device_id" $ fromString "Playing on "
  select ! A.id "device_id" ! A.name "device_id" $ do
    for_ spotifyDevicesDevices $ \SpotifyDevice {..} ->
      let setSelected = if spotifyDeviceIsActive
            then (! A.selected (fromString spotifyDeviceId))
            else P.id
      in  (setSelected $ option ! A.value (fromString spotifyDeviceId))
            (toHtml spotifyDeviceName)

empty = toHtml ("" :: String)


SpotifyConnect.hs
#! /usr/bin/env stack
-- stack script --resolver lts-14.16 --package lens --package scotty --package http-client --package http-types --package random --package http-conduit --package text --package aeson --package base64-bytestring --package bytestring --package lens-aeson --package MonadRandom

{-# LANGUAGE DeriveGeneric       #-}
{-# LANGUAGE OverloadedStrings   #-}
{-# LANGUAGE RecordWildCards     #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TypeApplications    #-}
import           Control.Lens
import           Control.Monad
import           Control.Monad.IO.Class
import           Control.Monad.Random
import           Data.Aeson
import qualified Data.Aeson                 as Aeson
import           Data.Aeson.Lens
import qualified Data.ByteString.Base64     as Base64
import qualified Data.ByteString.Char8      as S8
import qualified Data.ByteString.Lazy.Char8 as LBS
import qualified Data.Text                  as T
import qualified Data.Text.Lazy             as TL
import           GHC.Generics
import           Network.HTTP.Simple
import           Network.HTTP.Types
import           System.Environment
import           System.Random
import           Web.Scotty
import qualified Web.Scotty                 as Scotty

stateKey = "spotify_auth_state"
redirectUri = "http://localhost:8888/callback"

main :: IO ()
main = do
    clientId <- getEnv "SPOTIFY_CLIENT_ID"
    clientSecret <- getEnv "SPOTIFY_CLIENT_SECRET"

    let
      token = Base64.encode (S8.pack clientId <> ":" <> S8.pack clientSecret)
      authorization = "Basic " <> token

    scotty 8888 $ do
      Scotty.get "/" $ do
        Scotty.html randomizeForm

      Scotty.get "/randomize" $ do
        state <- liftIO $ fmap S8.pack $ replicateM 16 $ randomRIO ('a', 'z')
        -- setCookie stateKey state

        let url = TL.pack $ S8.unpack $ ("https://accounts.spotify.com/authorize?" <>) $
               mconcat
              [ "response_type=code"
              , "&client_id=" <> S8.pack clientId
              , "&scope=user-read-private user-read-email user-read-playback-state user-modify-playback-state user-library-read"
              , "&redirect_uri=http://localhost:8888/callback"
              , "&state=" <> state
              ]
        redirect url

      Scotty.get "/callback" $ do
        code <- param "code"

        -- _mState <- param "state"
        -- NB. this is important in prod but who cares here
        -- mStoredState <- (readCookie stateKey) =<< getCookies
        -- when mState != mStoredState $
          -- error "Mismatched state"
        -- clearCookie stateKey

        request0 <- parseRequest "POST https://accounts.spotify.com/api/token"

        let
          request
            = setRequestBodyURLEncoded
              [ ("grant_type", "authorization_code")
              , ("code", code)
              , ("redirect_uri", redirectUri)
              ]
            $ setRequestHeader "Authorization" [authorization] request0

        tokenResponse <- httpJSON request

        -- TODO: Store tokens somewhere?
        let
          tokenResponseValue = getResponseBody @Value tokenResponse
          Just accessToken = tokenResponseValue ^? key "access_token" . _String
          Just refreshToken = tokenResponseValue ^? key "refresh_token" . _String

        -- res <- liftIO $ callSpotifyWith accessToken "GET https://api.spotify.com/v1/me" id
        -- TODO: Choose a Device?
        -- res :: SpotifyDevices <- liftIO
          -- $ callSpotifyWith accessToken "GET https://api.spotify.com/v1/me/player/devices" id
        albums <- liftIO $ getMyAlbums accessToken
        randomAlbum <- liftIO $ uniform albums

        liftIO $ print $ length albums

        let
          tracks = map spotifyTrackUri $ spotifyTracksItems . spotifyAlbumTracks . spotifyAlbumItemAlbum $ randomAlbum
        liftIO $ callSpotifyWithNoResponse accessToken "PUT https://api.spotify.com/v1/me/player/play"
          $ setRequestBodyJSON (object ["uris" Aeson..= tracks])

        Scotty.html $ TL.unlines
          [ "<h1>Playing album: " <> TL.pack (spotifyAlbumName $ spotifyAlbumItemAlbum randomAlbum) <> "</h1>"
          , randomizeForm
          ]

randomizeForm = "<form action=\"/randomize\"><button>Randomize Album</button><form>"

getMyAlbums accessToken = go [] "https://api.spotify.com/v1/me/albums?limit=50"
 where
  go albums url = do
    SpotifyAlbumsResponse{..} <- callSpotifyWith accessToken ("GET " <> url) id
    case spotifyAlbumsResponseNext of
      Nothing   -> pure albums
      Just next -> do
        print next
        (spotifyAlbumsResponseItems ++) <$> go albums next

data SpotifyDevices = SpotifyDevices
  { spotifyDevicesDevices       :: [SpotifyDevice]
  } deriving (Show, Generic)

instance FromJSON SpotifyDevices where
  parseJSON = genericParseJSON (unPrefix "spotifyDevices")

instance ToJSON SpotifyDevices where
  toJSON = genericToJSON (unPrefix "spotifyDevices")
  toEncoding = genericToEncoding (unPrefix "spotifyDevices")

data SpotifyDevice = SpotifyDevice
  { spotifyDeviceId               :: String
  , spotifyDeviceIsActive         :: Bool
  , spotifyDeviceName             :: String
  , spotifyDeviceIsPrivateSession :: Bool
  , spotifyDeviceIsRestricted     :: Bool
  , spotifyDeviceVolumePercent    :: Double
  } deriving (Show, Generic)

instance FromJSON SpotifyDevice where
  parseJSON = genericParseJSON (unPrefix "spotifyDevice")

instance ToJSON SpotifyDevice where
  toJSON = genericToJSON (unPrefix "spotifyDevice")
  toEncoding = genericToEncoding (unPrefix "spotifyDevice")

unPrefix str = defaultOptions { fieldLabelModifier = camelTo2 '_' . drop (length (str :: String)) }

data SpotifyAlbumsResponse = SpotifyAlbumsResponse
  { spotifyAlbumsResponseItems :: [SpotifyAlbumItem]
  , spotifyAlbumsResponseNext  :: Maybe String
  } deriving (Show, Generic)

instance ToJSON SpotifyAlbumsResponse where
  toJSON = genericToJSON $ unPrefix "spotifyAlbumsResponse"
instance FromJSON SpotifyAlbumsResponse where
  parseJSON = genericParseJSON $ unPrefix "spotifyAlbumsResponse"

data SpotifyAlbumItem = SpotifyAlbumItem
  { spotifyAlbumItemAlbum :: SpotifyAlbum
  } deriving (Show, Generic)

instance ToJSON SpotifyAlbumItem where
  toJSON = genericToJSON $ unPrefix "spotifyAlbumItem"
instance FromJSON SpotifyAlbumItem where
  parseJSON = genericParseJSON $ unPrefix "spotifyAlbumItem"

data SpotifyAlbum = SpotifyAlbum
  { spotifyAlbumUri    :: String
  , spotifyAlbumName   :: String
  , spotifyAlbumTracks :: SpotifyTracks
  } deriving (Show, Generic)

instance ToJSON SpotifyAlbum where
  toJSON = genericToJSON $ unPrefix "spotifyAlbum"
instance FromJSON SpotifyAlbum where
  parseJSON = genericParseJSON $ unPrefix "spotifyAlbum"

data SpotifyTracks = SpotifyTracks { spotifyTracksItems :: [SpotifyTrack], spotifyTracksNext  :: Maybe String }
  deriving (Show, Generic)

instance ToJSON SpotifyTracks where
  toJSON = genericToJSON $ unPrefix "spotifyTracks"
instance FromJSON SpotifyTracks where
  parseJSON = genericParseJSON $ unPrefix "spotifyTracks"

data SpotifyTrack = SpotifyTrack { spotifyTrackUri :: String }
  deriving (Show, Generic)

instance ToJSON SpotifyTrack where
  toJSON = genericToJSON $ unPrefix "spotifyTrack"
instance FromJSON SpotifyTrack where
  parseJSON = genericParseJSON $ unPrefix "spotifyTrack"

-- {
  -- "href" : "https://api.spotify.com/v1/me/albums?offset=0&limit=1",
  -- "items" : [ {
    -- "added_at" : "2015-11-26T19:13:31Z",
    -- "album" : {
      -- "album_type" : "album",
      -- "artists" : [ {} ],
      -- "id" : "5m4VYOPoIpkV0XgOiRKkWC",
      -- "name" : "In & ut",
      -- "tracks" : {
        -- "href" : "https://api.spotify.com/v1/albums/5m4VYOPoIpkV0XgOiRKkWC/tracks?offset=0&limit=50",
        -- "items" : [ {
          -- "artists" : [ {
            -- "external_urls" : {
              -- "spotify" : "https://open.spotify.com/artist/58RMTlPJKbmpmVk1AmRK3h"
            -- },
            -- "href" : "https://api.spotify.com/v1/artists/58RMTlPJKbmpmVk1AmRK3h",
            -- "id" : "58RMTlPJKbmpmVk1AmRK3h",
            -- "name" : "Abidaz",
            -- "type" : "artist",
            -- "uri" : "spotify:artist:58RMTlPJKbmpmVk1AmRK3h"
          -- }, {
            -- "external_urls" : {
              -- "spotify" : "https://open.spotify.com/artist/1l63szZeUpN1m87MOD1u7K"
            -- },
            -- "href" : "https://api.spotify.com/v1/artists/1l63szZeUpN1m87MOD1u7K",
            -- "id" : "1l63szZeUpN1m87MOD1u7K",
            -- "name" : "Chapee",
            -- "type" : "artist",
            -- "uri" : "spotify:artist:1l63szZeUpN1m87MOD1u7K"
          -- }, {
            -- "external_urls" : {
              -- "spotify" : "https://open.spotify.com/artist/1VLf7Ncxb5Jga6eyd3jh6K"
            -- },
            -- "href" : "https://api.spotify.com/v1/artists/1VLf7Ncxb5Jga6eyd3jh6K",
            -- "id" : "1VLf7Ncxb5Jga6eyd3jh6K",
            -- "name" : "C.U.P",
            -- "type" : "artist",
            -- "uri" : "spotify:artist:1VLf7Ncxb5Jga6eyd3jh6K"
          -- } ],
          -- "available_markets" : [ "AR", "AT", "AU", "BE", "BR", "CL", "CO", "CY", "CZ", "DE" ],
          -- "disc_number" : 1,
          -- "duration_ms" : 170920,
          -- "explicit" : false,
          -- "external_urls" : {
            -- "spotify" : "https://open.spotify.com/track/3VNWq8rTnQG6fM1eldSpZ0"
          -- },
          -- "href" : "https://api.spotify.com/v1/tracks/3VNWq8rTnQG6fM1eldSpZ0",
          -- "id" : "3VNWq8rTnQG6fM1eldSpZ0",
          -- "name" : "E.C.",
          -- "preview_url" : "https://p.scdn.co/mp3-preview/f95e0dba1a76b44fa2b52da2bc273d4f1c4126a5",
          -- "track_number" : 1,
          -- "type" : "track",
          -- "uri" : "spotify:track:3VNWq8rTnQG6fM1eldSpZ0"
        -- }, {
          -- ...
        -- }, {
          -- "artists" : [ {
            -- "external_urls" : {
              -- "spotify" : "https://open.spotify.com/artist/58RMTlPJKbmpmVk1AmRK3h"
            -- },
            -- "href" : "https://api.spotify.com/v1/artists/58RMTlPJKbmpmVk1AmRK3h",
            -- "id" : "58RMTlPJKbmpmVk1AmRK3h",
            -- "name" : "Abidaz",
            -- "type" : "artist",
            -- "uri" : "spotify:artist:58RMTlPJKbmpmVk1AmRK3h"
          -- } ],
          -- "available_markets" : [ "AR", "AT", "AU", "BE", "BR", "CL", "CO", "CY", "CZ", "DE", "DK", "EE" ],
          -- "disc_number" : 1,
          -- "duration_ms" : 165946,
          -- "explicit" : false,
          -- "external_urls" : {
            -- "spotify" : "https://open.spotify.com/track/6ZrVKylVlxkaXHj42O0q2r"
          -- },
          -- "href" : "https://api.spotify.com/v1/tracks/6ZrVKylVlxkaXHj42O0q2r",
          -- "id" : "6ZrVKylVlxkaXHj42O0q2r",
          -- "name" : "Råknas - Radio Edit",
          -- "preview_url" : "https://p.scdn.co/mp3-preview/a7c9a4bfa9e346e3733e9d88076ad1ae409136fb",
          -- "track_number" : 13,
          -- "type" : "track",
          -- "uri" : "spotify:track:6ZrVKylVlxkaXHj42O0q2r"
        -- } ],
        -- "limit" : 50,
        -- "next" : null,
        -- "offset" : 0,
        -- "previous" : null,
        -- "total" : 13
      -- },
      -- "type" : "album",
      -- "uri" : "spotify:album:5m4VYOPoIpkV0XgOiRKkWC"
    -- }
  -- } ],
  -- "limit" : 1,
  -- "next" : "https://api.spotify.com/v1/me/albums?offset=1&limit=1",
  -- "offset" : 0,
  -- "previous" : null,
  -- "total" : 19
-- }

callSpotifyWith :: forall a. (FromJSON a, Show a) => T.Text ->  String -> (Request -> Request) -> IO a
callSpotifyWith accessToken request f = do
  request0 <- parseRequest request

  let
    spotifyAuthorization = "Bearer " <> S8.pack (T.unpack accessToken)
    request
      = setRequestHeader "Authorization" [spotifyAuthorization]
      $ setRequestHeader "Content-Type" ["application/json"]
      $ f request0

  response <- httpJSON request
  pure $ getResponseBody @a response

callSpotifyWithNoResponse :: T.Text ->  String -> (Request -> Request) -> IO ()
callSpotifyWithNoResponse accessToken request f = do
  request0 <- parseRequest request

  let
    spotifyAuthorization = "Bearer " <> S8.pack (T.unpack accessToken)
    request
      = setRequestHeader "Authorization" [spotifyAuthorization]
      $ setRequestHeader "Content-Type" ["application/json"]
      $ f request0

  response <- httpNoBody request
  if getResponseStatusCode response == 204 then pure () else do
    response1 <- httpJSON request
    print (getResponseBody response1 :: Value)
    error (show response1)

